### PR TITLE
[WIP] Performance improvements

### PR DESCRIPTION
**Pull request by Bastian**
There are some old code parts, that have a very bad performance (e.g. by too excessive use of streams). This PR tries to clean them up.

### CODE REVIEW

File: javacord-api/src/main/java/org/javacord/api/DiscordApi.java [hunk 1]
```diff
@@ -951,19 +952,26 @@ default AccountUpdater createAccountUpdater() {
...
     default Collection<PrivateChannel> getPrivateChannels() {
+        // TODO bad
         return Collections.unmodifiableList(
-                getCachedUsers().stream()
-                        .map(User::getPrivateChannel)
-                        .filter(Optional::isPresent)
-                        .map(Optional::get)
+                getChannels().stream()
+                        .filter(channel -> channel.getType() == ChannelType.PRIVATE_CHANNEL)
+                        .map(channel -> (PrivateChannel) channel)
```
Comment (burdoto) [inline javacord-api/src/main/java/org/javacord/api/DiscordApi.java]: Wouldn't this call be better with `PrivateChannel.class::cast`?

Comment (Bastian) [reply javacord-api/src/main/java/org/javacord/api/DiscordApi.java]: Yes 👍

File: javacord-api/src/main/java/org/javacord/api/DiscordApi.java [hunk 2]
```diff
@@ -1017,11 +1023,16 @@ default AccountUpdater createAccountUpdater() {
...
     default Collection<TextChannel> getTextChannels() {
-        Collection<TextChannel> channels = new ArrayList<>();
-        channels.addAll(getPrivateChannels());
-        channels.addAll(getServerTextChannels());
-        channels.addAll(getGroupChannels());
-        return Collections.unmodifiableCollection(channels);
+        // TODO bad
+        return Collections.unmodifiableList(
+                getChannels().stream()
+                        .filter(channel ->
```
Comment (Saladoc) [inline javacord-api/src/main/java/org/javacord/api/DiscordApi.java]: `.filter(TextChannel.class::instanceOf)`?

File: javacord-api/src/main/java/org/javacord/api/DiscordApi.java [hunk 3]
```diff
@@ -1030,11 +1041,16 @@ default AccountUpdater createAccountUpdater() {
...
     default Collection<VoiceChannel> getVoiceChannels() {
-        Collection<VoiceChannel> channels = new ArrayList<>();
-        channels.addAll(getPrivateChannels());
-        channels.addAll(getServerVoiceChannels());
-        channels.addAll(getGroupChannels());
-        return Collections.unmodifiableCollection(channels);
+        // TODO bad
+        return Collections.unmodifiableList(
+                getChannels().stream()
+                        .filter(channel ->
```
Comment (Saladoc) [inline javacord-api/src/main/java/org/javacord/api/DiscordApi.java]: `.filter(VoiceChannel.class::instanceOf)`?

**Comment by coaxial1** [comment]
Some exceptions that pop up due to the changes:

```
> 2019-02-17 01:11:24.662  WARN 27255 --- [dlers Processor] o.j.core.util.gateway.PacketHandler      : Couldn't handle packet of type CHANNEL_DELETE. Please contact the developer! (packet: {"user_limit":2,"type":2,"position":10,"permission_overwrites":[],"parent_id":"528369918241013811","nsfw":false,"name":"║Fortnite - Duo [smply | SmileX]","id":"546483632252911648","guild_id":"528368817924079617","bitrate":64000})
> 
> java.lang.ClassCastException: class org.javacord.core.entity.channel.ServerVoiceChannelImpl cannot be cast to class org.javacord.core.util.Cleanupable (org.javacord.core.entity.channel.ServerVoiceChannelImpl and org.javacord.core.util.Cleanupable are in unnamed module of loader org.springframework.boot.loader.LaunchedURLClassLoader @61df66b6)
>         at org.javacord.core.DiscordApiImpl.lambda$removeChannelFromCache$15(DiscordApiImpl.java:680) ~[javacord-3.0.3.jar!/:na]
>         at java.base/java.util.concurrent.ConcurrentHashMap.computeIfPresent(ConcurrentHashMap.java:1822) ~[na:na]
>         at org.javacord.core.DiscordApiImpl.removeChannelFromCache(DiscordApiImpl.java:679) ~[javacord-3.0.3.jar!/:na]
>         at org.javacord.core.util.handler.channel.ChannelDeleteHandler.handle(ChannelDeleteHandler.java:45) ~[javacord-3.0.3.jar!/:na]
>         at org.javacord.core.util.gateway.PacketHandler.lambda$handlePacket$0(PacketHandler.java:52) ~[javacord-3.0.3.jar!/:na]
>         at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515) ~[na:na]
>         at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264) ~[na:na]
>         at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) ~[na:na]
>         at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) ~[na:na]
>         at java.base/java.lang.Thread.run(Thread.java:834) ~[na:na]
```


And another one
```
2019-02-17 01:50:30.324  WARN 32654 --- [dlers Processor] o.j.core.util.gateway.PacketHandler      : Couldn't handle packet of type CHANNEL_DELETE. Please contact the developer! (packet: {"user_limit":2,"type":2
,"position":46,"permission_overwrites":[{"type":"role","id":"448599107301408768","deny":262208,"allow":1051648},{"type":"role","id":"463442484765392897","deny":0,"allow":263232},{"type":"role","id":"448599949811
253259","deny":0,"allow":268700752},{"type":"role","id":"516360777376989185","deny":0,"allow":268700752},{"type":"role","id":"535592865380696064","deny":0,"allow":1051648},{"type":"role","id":"470323386468270081
","deny":0,"allow":265280},{"type":"role","id":"470323385143001098","deny":0,"allow":263232},{"type":"role","id":"448600019252150283","deny":0,"allow":263232},{"type":"role","id":"454040113945640976","deny":0,"a
llow":268700752}],"parent_id":"543285593829015553","nsfw":false,"name":"👊 Duo PP #1","id":"546493392562159616","guild_id":"448599107301408768","bitrate":64000})

java.lang.ClassCastException: class org.javacord.core.entity.channel.ServerVoiceChannelImpl cannot be cast to class org.javacord.core.util.Cleanupable (org.javacord.core.entity.channel.ServerVoiceChannelImpl and
 org.javacord.core.util.Cleanupable are in unnamed module of loader org.springframework.boot.loader.LaunchedURLClassLoader @50eac852)
        at org.javacord.core.DiscordApiImpl.lambda$removeChannelFromCache$15(DiscordApiImpl.java:680) ~[javacord-3.0.3.jar!/:na]
        at java.base/java.util.concurrent.ConcurrentHashMap.computeIfPresent(ConcurrentHashMap.java:1822) ~[na:na]
        at org.javacord.core.DiscordApiImpl.removeChannelFromCache(DiscordApiImpl.java:679) ~[javacord-3.0.3.jar!/:na]
        at org.javacord.core.util.handler.channel.ChannelDeleteHandler.handle(ChannelDeleteHandler.java:45) ~[javacord-3.0.3.jar!/:na]
        at org.javacord.core.util.gateway.PacketHandler.lambda$handlePacket$0(PacketHandler.java:52) ~[javacord-3.0.3.jar!/:na]
        at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515) ~[na:na]
        at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264) ~[na:na]
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) ~[na:na]
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) ~[na:na]
        at java.base/java.lang.Thread.run(Thread.java:834) ~[na:na]
```

**Comment by Bastian** [comment]
All changes in this PR have been interpolated into other PRs.

**Pull request closed without merging**